app/models/shipit/user.rb

```diff
@@ -97,10 +97,16 @@ def github_user=(github_user)
         github_user = github_user.rels[:self].get.data
       end
 
+      github_email = if org_email?(github_user.email)
+        github_user.email
+      else
+        filter_org_emails(github_api.emails)
```

Comment:
I'm not super kin on that email selection code, can't really put my finger on it but something rubs me the wrong way.

I think casting the Octokit response to a simple array of strings would make for s much simpler code:

```ruby
github_email = github_user.email if org_email?(github_user.email)
github_email ||= github_api.emails.map(&:email).find { |e| org_email?(e) }
github_email ||= github_user.email
```

Reply:
Part of the complexity is from selecting the email marked as Primary if multiple valid org emails are found, and otherwise picking any of them if none are marked primary.
I could not think of a cleaner way of doing so.

Reply:
> Part of the complexity is from selecting the email marked as Primary if multiple valid org emails are found

But the primary is accessible as `user.email`, so if it's a valid org email, it's already set prior to entering that sub routine.

Reply:
If they do have a public email, then it should be guaranteed to be their primary one, so if we've specified a disqualifying org email, then it's as simple as selecting the first/any result from github_api.emails that fits the org_email.

The case I'm concerned about is if the user has no public email (e.g. github_user.email will be blank), and multiple private org emails, with one of those private emails marked as primary, and no guarantee on the primary email's position in the set returned by `github_api.emails`.

Perhaps I'm misunderstanding what you're saying, though.

Reply:
Ok, I didn't think of private emails. In this case:

```ruby
github_api.emails.sort_by { |e| e.primary ? 0 : 1 }.map(&:email).find { |e| org_email?(e) }
```

Also BTW, accessing private emails requires a specific permission: https://developer.github.com/v3/users/emails/#list-email-addresses-for-a-user `user:email`, does Shipit already ask for it, or do we have to add it?

Reply:
Ah! Good catch.

I was assuming our apps already had User permissions as they were already pulling User information in the first place, but it turns out we're only accessing the public user information through this API and currently have no user permissions enabled for our apps (Shopify Shipit, Shipit Mobile Staging and Shipit Mobile I think - let me know if I've missed any.)

I believe we'd just need to enable the permission (`User Permissions -> Email Addresses -> Read Only`) on the apps.

Based on the [app user permissions documentation](https://developer.github.com/apps/building-github-apps/identifying-and-authorizing-users-for-github-apps/#user-level-permissions), it's not clear to me whether adding this permission would be graceful, or if we'd suddenly see a bunch of permission access prompts for our org users.

I don't think we have user-to-server tokens to worry about, but I could be mistaken.

@DazWorrall any thoughts?